### DISCUSSION TITLE
livewire nesting component only works once

### DISCUSSION THREAD

**Discussion by amirmsj1374** [livewire/livewire#1895] [4. General]
Im using livewire component add product to shopping cart and change their quantities, there are three components.
product-quantity used as child components in table of cart component, when add product to cart by barcode, it only works for once and for more time it doesn't work and the quantiry of product in child component dosen't change, Also $refresh dosen't work too. how can I fix it? 

- Barcode.php

```
class Barcode extends Component
{
    public $barcode = null;
    public function addToCart()
    {
        $barcode = $this->barcode;

        $product = Product::where('barcode', $barcode)->where('shop_id', auth()->user()->shop_id)->first();

        //add the product to cart by barcode
        if ($product && $product->active) {
            $cart = \Cart::session(auth()->id())->add(array(
                'id' => $product->id,
                'name' => $product->title,
                'price' => $product->inventory->price_in_person,
                'quantity' => 1,
                'attributes' => array(
                    'color' => $product->color_code ?? '',
                    'discount' => $product->discount ? ($product->discount->discount_value * $product->inventory->price_in_person) : 0
                ),
                'associatedModel' => $product
            ));
        }

        $this->barcode = null;
       //emit to Cart.php
        $this->emit('cartUpdated');
    }

    public function render()
    {
        return view('livewire.barcode');
    }
}
```
-  and barcod.blade.php
```
<div>
    <input type="text" id="barcode" name="barcode" wire:model="barcode" wire:change="addToCart">
</div>
```

- Cart.php 
```
class Cart extends Component
{
    public $cartItems = [];

    protected $listeners = ['cartUpdated' => 'onCartUpdate'];
    public function mount()
    {
        $this->cartItems = \Cart::session(auth()->id())->getContent()->toArray();
    }

    public function deleteItemCart($itemId)
    {
        $cartItems = \Cart::session(auth()->id())->remove($itemId);
        $this->mount();
    }

    public function onCartUpdate()
    {
        $this->mount();
    }
    public function render()
    {
        $this->cartItems = \Cart::session(auth()->id())->getContent()->toArray();
        return view('livewire.cart');
    }
}
```
- cart.blade.php including a table

```
 <table class="table table-hover-animation mb-0">
    <thead>
        <tr>
            <th scope="col">product</th>
            <th scope="col">quantity</th>
            <th scope="col">price</th>
            <th scope="col">quantity from child</th>
            <th scope="col"></th>
        </tr>
    </thead>
    <tbody>
        @foreach ($cartItems as $item)
        <tr>
            <td>{{$item['name']}}</td>
             <td>{{$item['quantity']}}</td>
             <td>{{Cart::session(auth()->id())->get($item['id'])->getPriceSum()}}</td>
             <td>
                 <livewire:mall-cart :item="$item" wire:key="$item['id']"/>
             </td>
             <td>
                 <a href="javascript:;" wire:click="deleteItemCart({{ $item['id'] }})">
                     <i class="fa fa-times-circle-o text-danger fa-2x"></i>
                 </a>
             </td>
         </tr>
         @endforeach
     </tbody>
 </table>
```
- ProductQuantity.php
```
class ProductQuantity extends Component
{
    public $item;
    public $quantity;

    public function mount($item)
    {
        $this->item = $item;
        $this->quantity = $item['quantity'];
    }

   //change quantity from quantity input
    public function updateCart()
    {
        \Cart::session(auth()->id())->update($this->item['id'], [
            'quantity' => array(
                'relative' => false,
                'value' => $this->quantity,
            ),
        ]);

        $this->emit('cartUpdated');
    }

    public function render()
    {
        return view('livewire.product-quantity');
    }
}
```
- product-quantity.blade.php
```
<div>
    <input type="number" wire:model="quantity" wire:change="updateCart">
</div>
```

**Comment by Majid-Kamali** [comment]
I have this problem too and i test with tag div but not working can any one help

**Comment by iksaku** [comment]
`wire:key` is used only on traditional HTML elements, or Blade Components. When using Livewire components, use `:key`

**Reply by amirmsj1374** [reply]
I replaced :key but the quantity in child component dose not change for second time, maybe its becuase that I passed $item to child component via mount($item).  if this is the reason, how can change the quantity of products in the table each time?
also I should to say the quantity changes when the page is refreshed.

**Reply by iksaku** [reply]
Remember that Livewire components are not reactive between each other. The only way at the moment that you can make them to communicate is by [appending data to your events](https://laravel-livewire.com/docs/2.x/events#passing-parameters) and work with that data on the listener.

**Comment by marcotas** [comment]
There's a trick that always work for me when I want to make a child Livewire component "reactive". In your parent livewire component you use a timestamp for the key attribute:
```php
<livewire:child-component key="{{ now() }}" :my-prop="$myModel" />
```
The tricky part is the `now()` function in `key` prop.
This component now will always be updated when the parent is updated.

**Reply by amirmsj1374** [reply]
Thank you  @marcoT89 , it was clever way and worked easily.
`key={{ now() }}`
:+1:

**Reply by marcotas** [reply]
You're welcome! It's a good idea to take a look at [this discussion](https://github.com/livewire/livewire/discussions/2097) as well.

**Reply by tvthombre** [reply]
Great !!! Thanks !!!

**Reply by tvthombre** [reply]
I need help on Checkboxes, which is unchecked for read & write permission eventhough data of 1 is available

E:\L9_Apps\WireApp\app\Http\Livewire\Form\EmployeeList.php
E:\L9_Apps\WireApp\resources\views\livewire\form\employee-list.blade.php
E:\L9_Apps\WireApp\app\Http\Livewire\Form\Permits.php
E:\L9_Apps\WireApp\resources\views\livewire\form\permits.blade.php
